### DISCUSSION TITLE
DBX: Is there a way to use FindRecordsByExpr with the expression having an array?

### DISCUSSION THREAD

**Discussion by timstr97** [pocketbase/pocketbase#4245] [Q&A] [answered]
Hello,

i have the following problem. I am currently building a custom endpoint for getting userdata (in this case for the role applicant). The one who wants to get the data is the role teacher - the teacher should get the applicant if the teacher and the applicant are in the same class. 

I have done this in the pocketbase API-Rules and it works, but it is very slow because the rule is very big and creates a slow sql. So i have to write my own endpoint. 

This is the classes collection:
![image](https://github.com/pocketbase/pocketbase/assets/26687030/134e37a7-2d25-4180-a8a0-1794072ebb8f)

This is the part of the applicant api-rule in pocketbase, where the teacher role should access the applicant (the rest of the rule is not important).

```
  (
    @request.auth.collectionName = "teachers" &&
    @collection.classes.applicants ?~ id &&
    @collection.classes.teachers ?~ @request.auth.id
  )
```

The problem is, that the applicants-fields and the teachers-field in the classes collection are arrays. So in my endpoint i tried writing a sql query (or trying other functions for getting the data) but i didn't get it to work.

### My attempts 

> I tried only getting the classes for the applicant first, before trying to check if the teacher is in it. Note: The applicantId is one single Id, i want to do this for exactly one applicant.

1. FindRecordsByExpr: Gives no results. But this is no suprise, as applicants is an array. Is there a function for doing this with an array?

```
classes, err := server.Dao().FindRecordsByExpr("classes", dbx.HashExp{"applicants": applicantId})
```

2. Raw sql: Gives no results but it works when i put the query in DB Browser for SQLite, why is that so? I know that Like Operations are not that fast, but i need to get it to work somehow 
```
type Class struct {
Id              string                  `db:"id" json:"id"`
Applicants      types.JsonArray[string] `db:"applicants" json:"applicants"`
Teachers        types.JsonArray[string] `db:"teachers" json:"teachers"`
School          string                  `db:"school" json:"school"`
Grade           string                  `db:"grade" json:"grade"`
ClassIdentifier string                  `db:"classIdentifier" json:"classIdentifier"`
}
classes := []Class{}

query := "SELECT * FROM classes WHERE applicants LIKE '%{:applicantId}%'"

err := server.Dao().DB()
.NewQuery(query)
.Bind(dbx.Params{"applicantId": applicantId,})
.All(&classes)
```

How can i achieve what i want? Thanks for your help in advance, i really appreciate it.

**Comment by ganigeorgiev** [comment]
> I have done this in the pocketbase API-Rules and it works, but it is very slow because the rule is very big and creates a slow sql.

You are not providing details to understand what is the "slow" part to attempt to optimize it.

In any case, the [SQLite json functions](https://www.sqlite.org/json1.html) should be used when working with serialized json data as in your case the array of ids.

> Gives no results but it works when i put the query in DB Browser for SQLite

The placeholder parameter doesn't seem correctly build and you can try wrapping the `%` symbols as part of the binded `applicationId` string.

**Reply by ganigeorgiev** [reply]
_Also note that you have to use dot-notation to access the id (or other field) of the multiple relation records, aka. try using `@collection.classes.applicants.id` and `@collection.classes.teachers.id`._

**Reply by timstr97** [reply]
Hey, thanks for the fast answer.

The slow part is the big API-Rule for the applicants collection and other collections where the applicant is related to (6). So when i want to get the full applicant data i need to make an indirect expand with 6 collections that all have the same API-rules (slightly different, `id` is `applicant` in the related collections api-rules)

The API-Rule:
```
// Public User with public token
(
  @request.auth.id = "" &&
  @request.query.publicToken ?= @collection.publicTokens.token && 
  @collection.publicTokens.applicant ?= id && 
  @request.query.publicToken != ""
) ||
// Authorized Users
@request.auth.id != "" &&
(
  // The applicant himself
  @request.auth.id = id ||
  // Employees
  (
    @request.auth.collectionName = "employees" &&
    @request.auth.business != "" &&
    @collection.applicantAccessRights.applicant ?= id &&
    (
      @collection.applicantAccessRights.canBeFound ?= "everybody" ||
      (
        @collection.applicantAccessRights.canBeFound ?= "onlyFollowed" &&
        @collection.applicantBusinessFavorites.applicant ?= id &&
        @collection.applicantBusinessFavorites.business ?= @request.auth.business
      ) ||
      (
        @collection.applicantAccessRights.canBeFound ?= "onlyApplied" &&
        @collection.applications.applicant ?= id &&
        @collection.applications.vacancy.business ?= @request.auth.business
      )
    )
  ) ||
  // Teachers
  (
    @request.auth.collectionName = "teachers" &&
    @collection.classes.applicants ?~ id &&
    @collection.classes.teachers ?~ @request.auth.id
  )
)
```

So what i've done is writing my own endpoint representing this rule that fetches all the collections for the applicant. This works good and is also very fast but i can't get the teacher rules right.

**Reply by ganigeorgiev** [reply]
> but i can't get the teacher rules right.

See the earlier comment https://github.com/pocketbase/pocketbase/discussions/4245#discussioncomment-8264284.

Note that you can also replace `?~` with `?=` for exact match.

**Reply by timstr97** [reply]
Thanks for the tips with the Pocketbase-Rule, i will implement it. 

But i still can't rely on the default pocketbase endpoints for this, because as i said, the performance is really slow when doing 6 indirect expand with this huge ruleset. I will need to do this over a costum endpoint because when i do a request like this will take 7 seconds to get finished. 

```
#pb.collection(Collections.Applicants)
      .getOne<ApplicantProfileData>(id,
      {
       expand: "hobbies(applicant), experiences(applicant), languages(applicant), qualifications(applicant), schoolVisits(applicant), attachments(applicant)",
       })
```

The endpoint (without the teacher rule) runs fast so far, 11ms. 

I will try using SQLite json functions for the teacher-part. Will i have to write a custom query in this case or can i use FindRecordsByExpression?

**Reply by ganigeorgiev** [reply, marked as answer]
I was going to try to recreate your collections based on the above API rule and the limited details you provided to try to test and inspect what actually is the slow part (eg. missing index(es) and or just poor performance of the many joins) but realize that it will depend on each dataset size and I really don't have time right now to invest in this.

In general the only optimizations that I can think of at the moment that could be applied are:
- remove the `json` normaliziations (aka. the `json_valid` checks)
- test the performance difference between join and subquery when using `json_each()`
- for the cases where only `.id` is used avoid joining the entire related table and just join the `json_each` of the array field

I'll try to evaluate the above sometime in the future but there are no ETAs when that will be done.

In any case, I still don't quite understand the full picture of what you are trying to do, but if you still want to go with the raw query approach, here are some possible SQL statements you could try:
```sql
-- may require adding DISTINCT when used with the JOIN statements
SELECT classes.* FROM classes

-- v1 LEFT JOIN with WHERE
LEFT JOIN json_each(applicants) as je
WHERE je.value = {:applicantId}

-- v2 INNER JOIN with ON (should be similar to the above)
INNER JOIN json_each(applicants) as je ON je.value = {:applicantId}

-- v3 IN/EXISTS with subquery
WHERE EXISTS (SELECT 1 FROM json_each(applicants) as je WHERE je.value = {:applicantId})
```

Also another alternative if you want to use the PocketBase filter syntax, could be to check the [`Dao.FindFirstRecordByFilter`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.21.1/daos#Dao.FindFirstRecordByFilter)/[`Dao.FindRecordsByFilter`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.21.1/daos#Dao.FindRecordsByFilter) helpers.

**Reply by timstr97** [reply]
Thanks for the help, i will give this a try! By the way, hugeeee thanks for pocketbase, It is really awesome to work with!
